src/procstat.c

```diff
@@ -254,6 +184,87 @@ static bool item_type_directory(struct procstat_item *item)
 	return (item->flags & STATS_ENTRY_FLAG_DIR);
 }
 
+/**
+ * @brief Find the mutex for a procstat item.
+ *
+ * Searches the mutex array for a mutex registered to the subtree root
+ * for the given item.  If one isn't found, registers a new mutex with
+ * either the global root or the subtree root.
+ *
+ * @param item the item to lock
+ *
+ * @return an initialized mutex
+ */
+static pthread_mutex_t *procstat_find_mutex(struct procstat_item *item) {
+	while (item->parent && item->parent->base.parent) {
```

Comment:
Here item->parent might be a stale pointer and cause a crash like we've seen last week.
This happens when an item with refcnt > 1 belongs to a directory we removed, and then the refcnt owner (fuse)
calls either fuse_release() or fuse_forget().
My point is we cannot assume an item has a parent when locking the item.
Why not just use an 8-bit mutex index inside a procstat_item?

Reply:
This is why real filesystems don't have any concept of "parent" or allow wholesale removal of subtrees. But I concede your point.

Reply:
Yep, we need each item to know which mutex it is protected with. The only place, I believe, where deal with a parent mutex that might be not the same one of the item (parent is root) is fuse_lookup and procstat_create/remove: in these cases we know the parent exists.